### PR TITLE
greenhills: add cmake support

### PR DESCRIPTION
**Pull request by extinguish**
## Summary
1. refactor the toolchain management in CMake
2. unify the cmake toolchain name style
3. support greenhills build procedure with CMake
4. add protect build for greenhills and gnu with CMake

## Impact
1. refactor the cmake toolchain management, each toolchain has its own cmake file, make the toolchain specific feature modification constrain in its own cmake file, which is more clear

## Testing

1. has pass the os test,
2. has test on gcc/ghs compiler

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -639,14 +610,14 @@ set(nuttx_libs ${nuttx_kernel_libs} ${nuttx_system_libs} ${nuttx_apps_libs}
 
 if(NOT CONFIG_ARCH_SIM)
 
-  # TODO: nostart/nodefault not applicable to nuttx toolchain
-  if(CONFIG_ARCH_TOOLCHAIN_TASKING)
```
Comment (anchao) [inline CMakeLists.txt:643]: Why did you delete so many recent changes? Did you directly overwrite the current version with your private code????

**Review by anchao — CHANGES_REQUESTED** [review]

File: arch/arm/src/cmake/armclang.cmake [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+                      $<$<COMPILE_LANGUAGE:CXX>:-fcheck-new>)
+endif()
+
+if(NOT CONFIG_CXX_RTTI)
+  add_compile_options($<$<COMPILE_LANGUAGE:CXX>:-fno-rtti>)
+endif()
+
+set(CMAKE_EXE_LINKER_FLAGS_INIT "--specs=nosys.specs")
+
+set(PREPROCES ${CMAKE_C_COMPILER} ${CMAKE_C_FLAG_ARGS} -E -P -x c)
```
Comment (anchao) [inline arch/arm/src/cmake/armclang.cmake:242]: ditto

File: CMakeLists.txt [hunk 3]
```diff
@@ -466,16 +469,7 @@ else()
...
-# Compiler options TODO: move elsewhere
-
-if("${CMAKE_C_COMPILER_ID}" STREQUAL "GNU")
-  if(CMAKE_C_COMPILER_VERSION VERSION_GREATER 4.9)
-    # force color for gcc > 4.9
-    add_compile_options(-fdiagnostics-color=always)
-  endif()
-endif()
-
-if(MSVC)
```
Comment (anchao) [inline CMakeLists.txt:478]: ditto

**Review by anchao — CHANGES_REQUESTED** [review]

File: CMakeLists.txt (hunk 3 above)
Comment (extinguish) [reply CMakeLists.txt:478]: this part has moved to ./arch/arm/src/cmake/gcc.cmake

File: CMakeLists.txt (hunk 1 above)
Comment (extinguish) [reply CMakeLists.txt:643]: > Why did you delete so many recent changes? Did you directly overwrite the current version with your private code????

Sorry for my mistake here, I will fix soon

File: cmake/nuttx_toolchain.cmake [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+      DEPENDS
+      REQUIRED
+      SOURCE_FILE
+      TARGET_FILE
+      ARGN
+      ${ARGN})
+
+    add_custom_command(
+      OUTPUT ${TARGET_FILE}
+      COMMAND ${PREPROCESS} -I${CMAKE_BINARY_DIR}/include ${SOURCE_FILE} >
```
Comment (xuxin930) [inline cmake/nuttx_toolchain.cmake:65]:
```suggestion
      COMMAND ${PREPROCESS} -I${CMAKE_BINARY_DIR}/include -I${NUTTX_CHIP_ABS_DIR}  ${SOURCE_FILE} >
```

let's CHIP_ABS_DIR by default too

**Comment by extinguish** [comment]
the current build break is fixed by: https://github.com/apache/nuttx/pull/13717
which is still under review

**Review by xiaoxiang781216 — APPROVED** [review]

**Review by xiaoxiang781216 — APPROVED** [review]

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by raiden00pl** [comment]
@extinguish @xiaoxiang781216 hi, after this PR the colored output from compilation with cmake doesn't work. 

Before this PR:

image

after this PR compiler output on terminal looks like normal text also for warnings and errors:

image

**Comment by xiaoxiang781216** [comment]
@extinguish @xuxin930 please take a look.

**Comment by extinguish** [comment]
> @extinguish @xuxin930 please take a look.

this is fixed is https://github.com/apache/nuttx/pull/14404, and is under review
